sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -113,23 +106,15 @@ trait OperationHelper extends AliasHelper with PredicateHelper {
 }
 
 /**
- * A pattern that matches any number of project or filter operations on top of another relational
- * operator.  All filter operators are collected and their conditions are broken up and returned
- * together with the top project operator.
- * [[org.apache.spark.sql.catalyst.expressions.Alias Aliases]] are in-lined/substituted if
- * necessary.
+ * A pattern that matches a plan node with only deterministic Project and/or Filter above it. It
+ * returns the input plan itself if there is no Project or Filter.
  */
-object PhysicalOperation extends OperationHelper with PredicateHelper {
-  override protected def legacyMode: Boolean = true
-}
-
-/**
- * A variant of [[PhysicalOperation]]. It matches any number of project or filter
- * operations even if they are non-deterministic, as long as they satisfy the
- * requirement of CollapseProject and CombineFilters.
- */
-object ScanOperation extends OperationHelper with PredicateHelper {
-  override protected def legacyMode: Boolean = false
+object NodeWithOnlyDeterministicProjectAndFilter {
+  def unapply(plan: LogicalPlan): Option[LogicalPlan] = plan match {
+    case Project(projectList, child) if projectList.forall(_.deterministic) => unapply(child)
+    case Filter(cond, child) if cond.deterministic => unapply(child)
+    case _ => Some(plan)
```

Comment:
Hmm, if it returns `Some` for all other cases, doesn't it mean it matches all input plans?

Reply:
Yes, and I explicitly call it out in the classdoc. This is the same in `PhysicalOperation`, which always returns `Some`, and the caller side will specifiy the expected type.